File: 2020/2020W48/PreppindataW48.py
import pandas as pd
import re

xlsx = pd.ExcelFile("F:/Data/2020W48 Input.xlsx")

stand = pd.read_excel(xlsx,'Stands')
stand = pd.concat([stand, 
                   pd.DataFrame([map(str.strip, x) \
                                 for x in stand['Accessed by Gates'].str.split(',').values.tolist()])],
                  axis=1, sort=False)
stand = stand.melt(id_vars=[c for c in stand.columns if not re.match('^\d', str(c))], value_name='Gate', var_name='To_drop').dropna()
stand['OneGate'] = stand['Accessed by Gates'].apply(lambda x: 'N' if re.search('(,)',x) else 'Y')
stand = pd.merge(stand.drop(['Accessed by Gates','To_drop'], axis=1), pd.read_excel(xlsx,'Remote Stands Accesibility'), on='Gate')
stand['Stand'] = stand['Stand'].apply(lambda x: int(re.search('(\d+)$',x).group(1)))
stand['Gate'] = stand['Gate'].apply(lambda x: int(re.search('(\d+)$',x).group(1)))
flight = pd.merge(stand, pd.read_excel(xlsx,'Stand Allocations 01.02.2020 AM', dtype={'Time':str}), on='Stand')
flight['Date Begin'] = flight['Time'].apply(lambda x: pd.to_datetime('01.02.2020.'+x, format='%d.%m.%Y.%H%M'))
flight['Date End'] = flight['Date Begin'].apply(lambda x: x+pd.Timedelta(minutes=45))
flight['Time factor'] = flight.apply(lambda x: x['Time to Reach Remote Stands'] * (-1 if x['Requires Bus?']=='Y' else 1), axis=1)
flight = flight.sort_values(['OneGate', 'Requires Bus?','Time factor','Flight'], ascending=[False, False, False, True])

gateall = pd.read_excel(xlsx,'Gate Availability')
gateall['Flight'],gateall['Stand'],gateall['Requires Bus?'],gateall['Time to Reach Stand'] = [None,None,None,None]
allocated = []
for _, row in flight.iterrows():
    if row['Flight'] in allocated:
        continue
    available = gateall[(gateall['Gate']==row['Gate']) & (gateall['Date']>=row['Date Begin']) & (gateall['Date']<row['Date End']) & pd.isnull(gateall['Flight'])]
    if len(available)==3:
        allocated.append(row['Flight'])
        gateall.loc[available.index,['Flight','Stand','Requires Bus?','Time to Reach Stand']] = \
            [row['Flight'],row['Stand'],row['Requires Bus?'],0 if row['Requires Bus?']=='N' else row['Time to Reach Remote Stands']]
    else:    
        continue

final = gateall[['Gate', 'Stand', 'Date', 'Flight', 'Requires Bus?', 'Time to Reach Stand']].copy()

File: 2020/2020W48/preppindataw48.R
library(dplyr)
library(tidyr)
library(readxl)
library(splitstackshape)
library(stringr)
library(fuzzyjoin)
library(lubridate)

xlsx <- "F:/Data/2020W48 Input.xlsx"

flight <- read_excel(xlsx, sheet = "Stands") %>%
  cSplit(., 'Accessed by Gates', ',') %>%
  pivot_longer(.,cols=starts_with('Accessed by Gates'), names_to=NULL, values_to='Gate', values_drop_na=TRUE) %>%
  merge(read_excel(xlsx, sheet = "Remote Stands Accesibility"), by='Gate') %>%
  group_by(`Stand`) %>%
  mutate('One Gate' = if_else(n()==1,'Y','N'),
         'Stand' = as.integer(str_extract(`Stand`,'\\d+')),
         'Gate' = as.integer(str_extract(`Gate`,'\\d+'))) %>%
  merge(read_excel(xlsx, sheet = "Stand Allocations 01.02.2020 AM"), by='Stand') %>%
  mutate('Date Begin' = as.POSIXct(paste('2020-02-01',`Time`),format='%Y-%m-%d %H%M', tz='UTC'),
         'Date End' =   `Date Begin` %m+% minutes(45),
         'Time Factor' = if_else(`Requires Bus?`=='Y',-1,1)*`Time to Reach Remote Stands`) %>%  
  arrange(., desc(`One Gate`), desc(`Requires Bus?`), desc(`Time Factor`), `Flight`) %>%
  select(-c('One Gate','Time','Time Factor'))

allc_cols <- c('Flight','Stand','Requires Bus?','Time to Reach Remote Stands') 
gateall <- read_excel(xlsx, sheet = "Gate Availability") %>%
  cbind(setNames(lapply(allc_cols, function(x) x=NA), allc_cols))

allocated <- NULL
for (row in 1:nrow(flight)) {
  if(flight[row, 'Flight'] %in% allocated) {
    next
  }
  aval <- gateall %>% filter(is.na(`Flight`) & `Gate`==flight[row, 'Gate'] &
                               `Date`>=flight[row, 'Date Begin'] & `Date`<flight[row, 'Date End'] ) %>% nrow(.)
  if(aval==3){
    gateall <- fuzzy_left_join(gateall, flight[row, ],
                               by = c('Gate','Date'='Date Begin','Date'='Date End'), match_fun = list(`==`,`>=`,`<`)) %>%
      mutate('Gate'=coalesce(`Gate.x`,`Gate.y`),
             'Flight' = coalesce(`Flight.x`, `Flight.y`),
             'Stand' = coalesce(`Stand.x`, `Stand.y`),
             'Requires Bus?' = coalesce(`Requires Bus?.x`, `Requires Bus?.y`),
             'Time to Reach Remote Stands' = coalesce(`Time to Reach Remote Stands.x`, `Time to Reach Remote Stands.y`)) %>%
      select(-matches('\\.x$|\\.y$|Date Begin|Date End'))
    
    allocated <- c(allocated, flight[row, 'Flight'])
  }else{
    next
  }
} 

final <- gateall %>%
  mutate('Time to Reach Stand' = if_else(`Requires Bus?`=='N',0,`Time to Reach Remote Stands`)) %>%
  select(c('Gate', 'Stand', 'Date', 'Flight', 'Requires Bus?', 'Time to Reach Stand'))

View(final)
